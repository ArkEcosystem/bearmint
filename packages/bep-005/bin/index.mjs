#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
