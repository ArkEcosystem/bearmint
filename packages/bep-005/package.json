{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
