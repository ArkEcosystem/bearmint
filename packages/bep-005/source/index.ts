export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
