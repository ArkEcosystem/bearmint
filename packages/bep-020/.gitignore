account
account_name
account_public_key
account_validator_address
application
milestone
tx
tx_receipt
tx_state
world
