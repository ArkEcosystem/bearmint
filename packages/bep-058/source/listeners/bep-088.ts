import type { Account, EventListener, StateStore } from '@bearmint/bep-013'
import { DoesNotSatisfyElectionConditionsError } from '@bearmint/bep-109'

export function makeBEP88Listener(): EventListener<{ account: Account; state: StateStore }> {
	return {
		async execute({ account, state }) {
			if (await state.getMultiStore().get('@bearmint/bep-058').has(account.address)) {
				throw new DoesNotSatisfyElectionConditionsError('resigned')
			}
		},
	}
}
