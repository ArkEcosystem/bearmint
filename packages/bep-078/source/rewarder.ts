import { bytesToHex, canonicalizeJson } from '@bearmint/bep-009'
import type {
	Cradle,
	FunctionReturning,
	ValidatorRewarder,
	ValidatorRewarderStrategy,
} from '@bearmint/bep-013'
import { abci } from '@bearmint/bep-018'
import { getModuleMilestone } from '@bearmint/bep-021'

/**
 * `getCandidateBlockProposer` can't be trusted to be the original proposer
 *
 * @see https://github.com/tendermint/tendermint/discussions/8326
 * @see https://github.com/cosmos/cosmos-sdk/blob/main/x/distribution/abci.go
 */
export function makeRewarder(cradle: Cradle): ValidatorRewarder {
	return {
		async execute(state) {
			const tendermintAddress: string = bytesToHex(state.getCandidateBlockProposer())

			if (!(await state.getAccountRepository().hasByValidatorAddress(tendermintAddress))) {
				throw new Error(`Validator (${tendermintAddress}) does not exist`)
			}

			const validator = await state.getAccountRepository().findByValidatorAddress(tendermintAddress)

			const reward = await cradle.Container.build(
				cradle.StrategyRegistry.get<FunctionReturning<ValidatorRewarderStrategy>>(
					'@bearmint/bep-078',
					getModuleMilestone<{
						strategy: string
					}>(cradle.ServiceProviderRepository, cradle.CommittedState, '@bearmint/bep-078').strategy,
				),
			).execute(state, validator, state.getMilestone().parameters.denominations.reward)

			await state.getAccountRepository().index([validator])

			return {
				events: [
					new abci.Event({
						attributes: [
							{
								key: 'address',
								value: validator.address,
							},
							{
								key: 'balances',
								value: JSON.stringify(canonicalizeJson(validator.balances)),
							},
						],
						type: 'validator.reward',
					}),
				],
				reward,
			}
		},
	}
}
