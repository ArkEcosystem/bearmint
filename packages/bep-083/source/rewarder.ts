import { assert, bytesToHex, canonicalizeJson } from '@bearmint/bep-009'
import type {
	Cradle,
	DelegatorRewarder,
	DelegatorRewarderStrategy,
	FunctionReturning,
} from '@bearmint/bep-013'
import { PROTO_OPTS } from '@bearmint/bep-013'
import { abci, BEP59Strategy } from '@bearmint/bep-018'
import { denom } from '@bearmint/bep-021'

/**
 * `getCandidateBlockProposer` can't be trusted to be the original proposer
 *
 * @see https://github.com/tendermint/tendermint/discussions/8326
 * @see https://github.com/cosmos/cosmos-sdk/blob/main/x/distribution/abci.go
 */
export function makeRewarder(cradle: Cradle): DelegatorRewarder {
	return {
		async execute(state, reward) {
			const validator = await state
				.getAccountRepository()
				.findByValidatorAddress(bytesToHex(state.getCandidateBlockProposer()))

			assert.defined<object>(validator.validator)

			const validatorState = await cradle.StakingServiceFactory.make(state).allByValidator(
				validator,
			)

			if (Object.keys(validatorState.delegators).length === 0) {
				return []
			}

			// Distribution Strategies are optional so we should exit if the validator has none. Otherwise we'll crash!
			if (await state.getMultiStore().get('@bearmint/bep-059').missing(validator.address)) {
				return []
			}

			const distributionStrategy = BEP59Strategy.fromBinary(
				await state.getMultiStore().get('@bearmint/bep-059').get(validator.address),
				PROTO_OPTS,
			)

			// Build the strategy
			const calculateShare = cradle.Container.build(
				cradle.StrategyRegistry.get<FunctionReturning<DelegatorRewarderStrategy>>(
					'@bearmint/bep-083',
					distributionStrategy.strategy,
				),
			)

			/* eslint-disable sort-keys-fix/sort-keys-fix */
			cradle.Logger.debug('%s', {
				module: '@bearmint/bep-083',
				action: 'execute',
				strategy: distributionStrategy.strategy,
			})
			/* eslint-enable sort-keys-fix/sort-keys-fix */

			const events: abci.Event[] = []

			for (const address of await cradle.StakingServiceFactory.make(state).keysByValidator(
				validator,
			)) {
				// No point in rewarding ourselves...
				if (address === validator.address) {
					continue
				}

				const delegator = await state.getAccountRepository().findByAddress(address)

				await calculateShare.execute(state, validator, delegator, reward, denom(state, 'reward'))

				await state.getAccountRepository().index([delegator, validator])

				events.push(
					new abci.Event({
						attributes: [
							new abci.EventAttribute({
								key: 'delegator.address',
								value: delegator.address,
							}),
							new abci.EventAttribute({
								key: 'delegator.balances',
								value: JSON.stringify(canonicalizeJson(delegator.balances)),
							}),
							new abci.EventAttribute({
								key: 'validator.address',
								value: validator.address,
							}),
							new abci.EventAttribute({
								key: 'validator.balances',
								value: JSON.stringify(canonicalizeJson(validator.balances)),
							}),
						],
						type: 'delegator.reward',
					}),
				)
			}

			return events
		},
	}
}
