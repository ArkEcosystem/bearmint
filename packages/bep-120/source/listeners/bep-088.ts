import type { Account, EventListener, StateStore } from '@bearmint/bep-013'
import { DoesNotSatisfyElectionConditionsError } from '@bearmint/bep-109'

import { isJailed } from '../utils.js'

export function makeBEP88Listener(): EventListener<{ account: Account; state: StateStore }> {
	return {
		async execute({ account, state }) {
			if (await isJailed(state, account)) {
				throw new DoesNotSatisfyElectionConditionsError('jailed')
			}
		},
	}
}
