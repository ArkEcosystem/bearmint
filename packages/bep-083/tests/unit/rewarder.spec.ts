import { expect, describe, it, beforeEach } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)
import { denominations } from '@bearmint/bep-006'
import {
	fakeAccountSerializer,
	fakeAddressFactory,
	fakeEventDispatcher,
	fakeMultiStore,
} from '@bearmint/bep-008'
import { hexToBytes } from '@bearmint/bep-009'
import type { Account, Container, StateStore, StrategyRegistry } from '@bearmint/bep-013'
import { ContainerType } from '@bearmint/bep-013'
import { makeContainer } from '@bearmint/bep-014'
import { BEP59Strategy, BEP88Validator } from '@bearmint/bep-018'
import { makeState } from '@bearmint/bep-020'
import { makeAccountRepository } from '@bearmint/bep-022'
import { makeStakingServiceFactory } from '@bearmint/bep-088'
import { makeStrategyRegistry } from '@bearmint/bep-146'

import { makeStrategy as BEP84 } from '../../../bep-084/distribution/index.js'
import { makeStrategy as BEP85 } from '../../../bep-085/distribution/index.js'
import { makeStrategy as BEP86 } from '../../../bep-086/distribution/index.js'
import { makeRewarder } from '../../source/rewarder.js'

const validatorAddress = '134a3ccc33350717a7db79a93b5516ea0a5d4db0a60f03a08335e0c4c975d15e'

describe<{
	container: Container
	delegator1: Account
	delegator2: Account
	state: StateStore
	strategyRegistry: StrategyRegistry
	validator: Account
}>('Delegator Rewarder', () => {
	beforeEach(async (context) => {
		context.container = makeContainer()
		context.state = await makeState(
			makeAccountRepository({
				AccountSerializer: fakeAccountSerializer(),
				AddressFactory: fakeAddressFactory(),
				EventDispatcher: fakeEventDispatcher(),
			}),
			await fakeMultiStore(),
		)
		context.strategyRegistry = makeStrategyRegistry()
		context.strategyRegistry.set('@bearmint/bep-083', '@bearmint/bep-084', BEP84)
		context.strategyRegistry.set('@bearmint/bep-083', '@bearmint/bep-085', BEP85)
		context.strategyRegistry.set('@bearmint/bep-083', '@bearmint/bep-086', BEP86)

		context.container.bindValue(ContainerType.Logger, console)
		context.container.bindValue(ContainerType.DeliverTxState, context.state)
		context.container.bindValue(
			ContainerType.StakingServiceFactory,
			makeStakingServiceFactory({ DataSink: { put() {} } }),
		)

		context.state.setMilestone({
			blockNumber: '1',
			parameters: {
				application: {},
				consensus: {},
				denominations,
				modules: { mandatory: {} },
				transactions: {},
			},
		})

		context.state.setCandidateBlock({ header: { proposerAddress: hexToBytes(validatorAddress) } })

		context.delegator1 = await fakeAccountSerializer().deserialize({
			address: 'bear1yenxkn0ln2se57zfsxprgua9fz3fvqwuj9twxq83vhvqtw6nn4ss6xhfdx',
		})
		context.delegator1.balances['GOLD'] = BigInt(0)
		context.delegator1.stakes['buckley'] = BigInt(80e8)

		context.delegator2 = await fakeAccountSerializer().deserialize({
			address: 'bear1zd9renpnx5r30f7m0x5nk4gkag996nds5c8s8gyrxhsvfjt4690qxdkwq3',
		})
		context.delegator2.balances['GOLD'] = BigInt(0)
		context.delegator2.stakes['buckley'] = BigInt(20e8)

		context.validator = await fakeAccountSerializer().deserialize({
			address: 'bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr',
			publicKey: validatorAddress,
		})
		context.validator.balances['GOLD'] = BigInt(2e8)
		context.validator.moniker = 'buckley'
		context.validator.validator = {
			address: validatorAddress,
		}

		await context.state.getMultiStore().add('@bearmint/bep-059')
		await context.state
			.getMultiStore()
			.get('@bearmint/bep-059')
			.set(
				context.validator.address,
				new BEP59Strategy({
					strategyArgs: '50',
					strategy: '@bearmint/bep-086',
				}).toBinary(),
			)

		await context.state.getMultiStore().add('@bearmint/bep-088')
		await context.state
			.getMultiStore()
			.get('@bearmint/bep-088')
			.set(
				context.validator.address,
				new BEP88Validator({
					address: validatorAddress,
					delegators: {
						[context.delegator1.address]: BigInt(80e8),
						[context.delegator2.address]: BigInt(20e8),
					},
				}).toBinary(),
			)

		await context.state
			.getAccountRepository()
			.index([context.delegator1, context.delegator2, context.validator])
	})

	it('should exit early if there are no delegators', async (context) => {
		await context.state
			.getMultiStore()
			.get('@bearmint/bep-088')
			.set(
				context.validator.address,
				new BEP88Validator({
					address: validatorAddress,
					delegators: {},
				}).toBinary(),
			)

		await context.state.getAccountRepository().index([context.validator])

		expect(
			await makeRewarder({
				Container: context.container,
				Logger: { debug: () => {} },
				StrategyRegistry: context.strategyRegistry,
				StakingServiceFactory: makeStakingServiceFactory({ DataSink: { put() {} } }),
			}).execute(context.state, BigInt(2e8)),
		).toHaveLength(0)
	})

	it('should throw if the strategy handler does not exist', async (context) => {
		context.validator.metadata = {
			'@bearmint/bep-059': {
				percentage: 50,
				strategy: 'borked',
			},
		}

		await context.state.getAccountRepository().index([context.validator])

		await expect(() =>
			makeRewarder({
				Container: context.container,
				Logger: { debug: () => {} },
				StrategyRegistry: context.strategyRegistry,
			}).execute(context.state, BigInt(2e8)),
		).rejects.toThrow()
	})

	it('should not distribute rewards when the mode is `standard`', async (context) => {
		context.validator.stakes[context.validator.moniker] = BigInt(80e8)

		await context.state
			.getMultiStore()
			.get('@bearmint/bep-088')
			.set(
				context.validator.address,
				new BEP88Validator({
					address: validatorAddress,
					delegators: {
						[context.validator.address]: BigInt(80e8),
					},
				}).toBinary(),
			)

		await context.state.getAccountRepository().index([context.validator])

		expect(context.delegator1.balances['GOLD'].toString()).toStrictEqual('0')
		expect(context.delegator2.balances['GOLD'].toString()).toStrictEqual('0')
		expect(context.validator.balances['GOLD'].toString()).toStrictEqual(`${2e8}`)

		expect(
			await makeRewarder({
				Container: context.container,
				Logger: { debug: () => {} },
				StrategyRegistry: context.strategyRegistry,
				StakingServiceFactory: makeStakingServiceFactory({ DataSink: { put() {} } }),
			}).execute(context.state, BigInt(2e8)),
		).toMatchSnapshot()

		expect(context.delegator1.balances['GOLD'].toString()).toStrictEqual('0')
		expect(context.delegator2.balances['GOLD'].toString()).toStrictEqual('0')
		expect(context.validator.balances['GOLD'].toString()).toStrictEqual(`${2e8}`)
	})

	it('should distribute rewards when the mode is `delegated`', async (context) => {
		expect(context.delegator1.balances['GOLD'].toString()).toStrictEqual('0')
		expect(context.delegator2.balances['GOLD'].toString()).toStrictEqual('0')
		expect(context.validator.balances['GOLD'].toString()).toStrictEqual(`${2e8}`)

		expect(
			await makeRewarder({
				Container: context.container,
				Logger: { debug: () => {} },
				StrategyRegistry: context.strategyRegistry,
				StakingServiceFactory: makeStakingServiceFactory({ DataSink: { put() {} } }),
			}).execute(context.state, BigInt(2e8)),
		).toMatchSnapshot()

		expect(
			(
				await context.state
					.getAccountRepository()
					.findByAddress('bear1yenxkn0ln2se57zfsxprgua9fz3fvqwuj9twxq83vhvqtw6nn4ss6xhfdx')
			).balances['GOLD'].toString(),
		).toStrictEqual('80000000')

		expect(
			(
				await context.state
					.getAccountRepository()
					.findByAddress('bear1zd9renpnx5r30f7m0x5nk4gkag996nds5c8s8gyrxhsvfjt4690qxdkwq3')
			).balances['GOLD'].toString(),
		).toStrictEqual('20000000')

		expect(
			(
				await context.state
					.getAccountRepository()
					.findByAddress('bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr')
			).balances['GOLD'].toString(),
		).toStrictEqual(`${1e8}`)
	})
})
