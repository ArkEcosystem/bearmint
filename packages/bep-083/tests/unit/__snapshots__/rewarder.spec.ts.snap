// Vitest Snapshot v1

exports[`Delegator Rewarder > should distribute rewards when the mode is \`delegated\` 1`] = `
[
  Event {
    "attributes": [
      EventAttribute {
        "index": false,
        "key": "delegator.address",
        "value": "bear1yenxkn0ln2se57zfsxprgua9fz3fvqwuj9twxq83vhvqtw6nn4ss6xhfdx",
      },
      EventAttribute {
        "index": false,
        "key": "delegator.balances",
        "value": "{\\"FIRE\\":\\"0\\",\\"GOLD\\":\\"80000000\\",\\"SEED\\":\\"0\\",\\"BEAR\\":\\"0\\"}",
      },
      EventAttribute {
        "index": false,
        "key": "validator.address",
        "value": "bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr",
      },
      EventAttribute {
        "index": false,
        "key": "validator.balances",
        "value": "{\\"FIRE\\":\\"0\\",\\"GOLD\\":\\"120000000\\",\\"SEED\\":\\"0\\",\\"BEAR\\":\\"0\\"}",
      },
    ],
    "type": "delegator.reward",
  },
  Event {
    "attributes": [
      EventAttribute {
        "index": false,
        "key": "delegator.address",
        "value": "bear1zd9renpnx5r30f7m0x5nk4gkag996nds5c8s8gyrxhsvfjt4690qxdkwq3",
      },
      EventAttribute {
        "index": false,
        "key": "delegator.balances",
        "value": "{\\"FIRE\\":\\"0\\",\\"GOLD\\":\\"20000000\\",\\"SEED\\":\\"0\\",\\"BEAR\\":\\"0\\"}",
      },
      EventAttribute {
        "index": false,
        "key": "validator.address",
        "value": "bear1vlezq7dtyml3zu3wgfdwdglcalradfnt5xpzahl7a5478hn5g20qxd5nkr",
      },
      EventAttribute {
        "index": false,
        "key": "validator.balances",
        "value": "{\\"FIRE\\":\\"0\\",\\"GOLD\\":\\"100000000\\",\\"SEED\\":\\"0\\",\\"BEAR\\":\\"0\\"}",
      },
    ],
    "type": "delegator.reward",
  },
]
`;

exports[`Delegator Rewarder > should not distribute rewards when the mode is \`standard\` 1`] = `[]`;
